### PR TITLE
feat: heatmap in tree view

### LINKED ISSUE #497: Heatmap/counts in tree view
### Is your feature request related to a problem? Please describe

It's not easily possible to identify elements in the tree view that might lack annotations.

### Desired solution

Color the entries in the tree view like a heatmap according to some counter and show this counter. The counter can be one of the following options:

* How often is the element used?
* How many annotations does the element have?

### Additional context (optional)

Heatmap: Relevant elements are red, less relevant elements are blue. The exact number should also be shown.
Heatmap should be able to be turned off and you should be able to choose from heatmaps (amount of uses, amount of annotations).

* Relates to #494.
* Is similar to #487.

### PR DESCRIPTION
**Pull request by lars-reimann**
Closes #497.

### Summary of Changes

Add heat map to tree view with two different modes:
* Usage mode: The fewer usages an API element has, the more blue the tag is. The more usages an API element has, the more red the tag becomes. The interpolation between the colors is logarithmic.
* Annotation mode: The fewer annotations an API element has, the more more its tag is. The more annotations it has, the more red the tag. The interpolation between the colors is linear.

#### Settings

image

#### Tags in Tree View

image

### Testing Instructions

1. Run this:
```sh
cd api-editor
./gradlew run
```
2. Load API, usage, and annotation data into the editor
3. Set the different heat map modes in the menu bar under "Settings"

### CODE REVIEW

File: api-editor/gui/src/features/packageData/treeView/TreeNode.tsx [hunk 1]
```diff
@@ -51,7 +66,9 @@ const TreeNode: React.FC<TreeNodeProps> = function ({ declaration, icon, isExpan
                 showChildren={showChildren}
                 isSelected={isSelected(declaration, currentPathname)}
             />
-            <Icon as={icon} />
+            {display_heat_map && <Box bg={heat_color} color="white" width={box_width} borderRadius="full" height="18px" fontWeight="bold"
```
Comment (lars-reimann) [inline api-editor/gui/src/features/packageData/treeView/TreeNode.tsx:69]: Can use a [tag](https://chakra-ui.com/docs/components/data-display/tag).

File: api-editor/gui/src/features/packageData/treeView/TreeNode.tsx [hunk 2]
```diff
@@ -65,4 +82,11 @@ const isSelected = function (declaration: PythonDeclaration, currentPathname: st
     return `/${declaration.pathAsString()}` === currentPathname;
 };
 
+const getColorFromValue = function (maxValue: number, specificValue: number): string {
+    const percentage = Math.log(specificValue) / Math.log(maxValue);
```
Comment (lars-reimann) [inline api-editor/gui/src/features/packageData/treeView/TreeNode.tsx:102]: Not quite correct:

|`specificValue`|`maxValue`|desired result|actual result|
|-|-|-|-|
|0|0|0|NaN|
|0|10|0|-Infinity|
|1|1|1|NaN|
|1|2|0<x<1|0|

Suggestion:

```text
If specificValue <= 0: return 0
Else if specificValue >= maxValue: return 1
Else: return log(specificValue + 1) / log(specificValue + 1)
```

The conditions `specificValue < 0` and `specificValue >= maxValue` are a precaution.

Comment (GideonKoenig) [reply api-editor/gui/src/features/packageData/treeView/TreeNode.tsx]: Has been implemented.
As a precaution the parameters are tested for impossible values.
Also the base values for the logarithmic calculation has been changed from 0 to 1, which should fix the third and fourth issue.

**Comment by lars-reimann** [comment]
The description of the PR (Summary of Changes, Screenshots, Testing Instructions) still needs to be updated. I'll have a look at the code tomorrow.

**Review by lars-reimann — COMMENTED** [review]
Switching between heat map modes crashes the app:

<details>
 <summary>Console output</summary>
    react-dom.development.js:15050 
        
       Uncaught Error: Rendered fewer hooks than expected. This may be caused by an accidental early return statement.
    at renderWithHooks (react-dom.development.js:15050:13)
    at updateFunctionComponent (react-dom.development.js:17356:20)
    at beginWork (react-dom.development.js:19063:16)
    at HTMLUnknownElement.callCallback2 (react-dom.development.js:3945:14)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:3994:16)
    at invokeGuardedCallback (react-dom.development.js:4056:31)
    at beginWork$1 (react-dom.development.js:23964:7)
    at performUnitOfWork (react-dom.development.js:22776:12)
    at workLoopSync (react-dom.development.js:22707:5)
    at renderRootSync (react-dom.development.js:22670:7)
    renderWithHooks @ react-dom.development.js:15050
    updateFunctionComponent @ react-dom.development.js:17356
    beginWork @ react-dom.development.js:19063
    callCallback2 @ react-dom.development.js:3945
    invokeGuardedCallbackDev @ react-dom.development.js:3994
    invokeGuardedCallback @ react-dom.development.js:4056
    beginWork$1 @ react-dom.development.js:23964
    performUnitOfWork @ react-dom.development.js:22776
    workLoopSync @ react-dom.development.js:22707
    renderRootSync @ react-dom.development.js:22670
    performSyncWorkOnRoot @ react-dom.development.js:22293
    (anonymous) @ react-dom.development.js:11327
    unstable_runWithPriority @ scheduler.development.js:468
    runWithPriority$1 @ react-dom.development.js:11276
    flushSyncCallbackQueueImpl @ react-dom.development.js:11322
    flushSyncCallbackQueue @ react-dom.development.js:11309
    discreteUpdates$1 @ react-dom.development.js:22420
    discreteUpdates @ react-dom.development.js:3756
    dispatchDiscreteEvent @ react-dom.development.js:5889
    6react_devtools_backend.js:4026 
      
       The above error occurred in the <FunctionNode> component:

    at FunctionNode (http://localhost:3000/src/features/packageData/treeView/FunctionNode.tsx?t=1654331351932:8:33)
    at div
    at http://localhost:3000/node_modules/.vite/@chakra-ui_react.js?v=4deaba1b:1635:45
    at http://localhost:3000/src/features/packageData/treeView/TreeView.tsx?t=1654331351932:63:35
    at div
    at div
    at List (http://localhost:3000/node_modules/.vite/react-window.js?v=4deaba1b:782:30)
    at div
    at AutoSizer2 (http://localhost:3000/node_modules/.vite/react-virtualized-auto-sizer.js?v=4deaba1b:288:5)
    at http://localhost:3000/src/features/packageData/treeView/TreeView.tsx?t=1654331351932:15:26
    at div
    at http://localhost:3000/node_modules/.vite/@chakra-ui_react.js?v=4deaba1b:1635:45
    at http://localhost:3000/node_modules/.vite/@chakra-ui_react.js?v=4deaba1b:19678:20
    at div
    at http://localhost:3000/node_modules/.vite/@chakra-ui_react.js?v=4deaba1b:1635:45
    at http://localhost:3000/node_modules/.vite/@chakra-ui_react.js?v=4deaba1b:19648:20
    at App (http://localhost:3000/src/app/App.tsx?t=1654331351932:49:20)
    at Router (http://localhost:3000/node_modules/.vite/chunk-3JTTZUC5.js?v=4deaba1b:433:15)
    at HashRouter (http://localhost:3000/node_modules/.vite/react-router-dom.js?v=4deaba1b:110:5)
    at EnvironmentProvider (http://localhost:3000/node_modules/.vite/@chakra-ui_react.js?v=4deaba1b:3307:24)
    at ColorModeProvider (http://localhost:3000/node_modules/.vite/@chakra-ui_react.js?v=4deaba1b:3469:21)
    at ThemeProvider2 (http://localhost:3000/node_modules/.vite/@chakra-ui_react.js?v=4deaba1b:1662:44)
    at ThemeProvider4 (http://localhost:3000/node_modules/.vite/@chakra-ui_react.js?v=4deaba1b:5415:27)
    at http://localhost:3000/node_modules/.vite/@chakra-ui_react.js?v=4deaba1b:2525:24
    at ChakraProvider2 (http://localhost:3000/node_modules/.vite/@chakra-ui_react.js?v=4deaba1b:5651:24)
    at Provider (http://localhost:3000/node_modules/.vite/react-redux.js?v=4deaba1b:368:20)

    Consider adding an error boundary to your tree to customize error handling behavior.
    Visit https://reactjs.org/link/error-boundaries to learn more about error boundaries.
    overrideMethod @ react_devtools_backend.js:4026
    logCapturedError @ react-dom.development.js:20085
    update.callback @ react-dom.development.js:20118
    callCallback @ react-dom.development.js:12318
    commitUpdateQueue @ react-dom.development.js:12339
    commitLifeCycles @ react-dom.development.js:20736
    commitLayoutEffects @ react-dom.development.js:23426
    callCallback2 @ react-dom.development.js:3945
    invokeGuardedCallbackDev @ react-dom.development.js:3994
    invokeGuardedCallback @ react-dom.development.js:4056
    commitRootImpl @ react-dom.development.js:23151
    unstable_runWithPriority @ scheduler.development.js:468
    runWithPriority$1 @ react-dom.development.js:11276
    commitRoot @ react-dom.development.js:22990
    performSyncWorkOnRoot @ react-dom.development.js:22329
    (anonymous) @ react-dom.development.js:11327
    unstable_runWithPriority @ scheduler.development.js:468
    runWithPriority$1 @ react-dom.development.js:11276
    flushSyncCallbackQueueImpl @ react-dom.development.js:11322
    flushSyncCallbackQueue @ react-dom.development.js:11309
    discreteUpdates$1 @ react-dom.development.js:22420
    discreteUpdates @ react-dom.development.js:3756
    dispatchDiscreteEvent @ react-dom.development.js:5889
    react-dom.development.js:15050 
       
       Uncaught Error: Rendered fewer hooks than expected. This may be caused by an accidental early return statement.
    at renderWithHooks (react-dom.development.js:15050:13)
    at updateFunctionComponent (react-dom.development.js:17356:20)
    at beginWork (react-dom.development.js:19063:16)
    at HTMLUnknownElement.callCallback2 (react-dom.development.js:3945:14)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:3994:16)
    at invokeGuardedCallback (react-dom.development.js:4056:31)
    at beginWork$1 (react-dom.development.js:23964:7)
    at performUnitOfWork (react-dom.development.js:22776:12)
    at workLoopSync (react-dom.development.js:22707:5)
    at renderRootSync (react-dom.development.js:22670:7)
    renderWithHooks @ react-dom.development.js:15050
    updateFunctionComponent @ react-dom.development.js:17356
    beginWork @ react-dom.development.js:19063
    callCallback2 @ react-dom.development.js:3945
    invokeGuardedCallbackDev @ react-dom.development.js:3994
    invokeGuardedCallback @ react-dom.development.js:4056
    beginWork$1 @ react-dom.development.js:23964
    performUnitOfWork @ react-dom.development.js:22776
    workLoopSync @ react-dom.development.js:22707
    renderRootSync @ react-dom.development.js:22670
    performSyncWorkOnRoot @ react-dom.development.js:22293
    (anonymous) @ react-dom.development.js:11327
    unstable_runWithPriority @ scheduler.development.js:468
    runWithPriority$1 @ react-dom.development.js:11276
    flushSyncCallbackQueueImpl @ react-dom.development.js:11322
    flushSyncCallbackQueue @ react-dom.development.js:11309
    discreteUpdates$1 @ react-dom.development.js:22420
    discreteUpdates @ react-dom.development.js:3756
    dispatchDiscreteEvent @ react-dom.development.js:5889
    react_devtools_backend.js:4026 
       
       Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.
    at App (http://localhost:3000/src/app/App.tsx?t=1654331351932:49:20)
    at Router (http://localhost:3000/node_modules/.vite/chunk-3JTTZUC5.js?v=4deaba1b:433:15)
    at HashRouter (http://localhost:3000/node_modules/.vite/react-router-dom.js?v=4deaba1b:110:5)
    at EnvironmentProvider (http://localhost:3000/node_modules/.vite/@chakra-ui_react.js?v=4deaba1b:3307:24)
    at ColorModeProvider (http://localhost:3000/node_modules/.vite/@chakra-ui_react.js?v=4deaba1b:3469:21)
    at ThemeProvider2 (http://localhost:3000/node_modules/.vite/@chakra-ui_react.js?v=4deaba1b:1662:44)
    at ThemeProvider4 (http://localhost:3000/node_modules/.vite/@chakra-ui_react.js?v=4deaba1b:5415:27)
    at http://localhost:3000/node_modules/.vite/@chakra-ui_react.js?v=4deaba1b:2525:24
    at ChakraProvider2 (http://localhost:3000/node_modules/.vite/@chakra-ui_react.js?v=4deaba1b:5651:24)
    at Provider (http://localhost:3000/node_modules/.vite/react-redux.js?v=4deaba1b:368:20)
    overrideMethod @ react_devtools_backend.js:4026
    printWarning @ react-dom.development.js:67
    error @ react-dom.development.js:43
    warnAboutUpdateOnUnmountedFiberInDEV @ react-dom.development.js:23914
    scheduleUpdateOnFiber @ react-dom.development.js:21840
    dispatchAction @ react-dom.development.js:16139
    getUsagesFromIndexedDB @ App.tsx:215
    await in getUsagesFromIndexedDB (async)
    (anonymous) @ App.tsx:70
    invokePassiveEffectCreate @ react-dom.development.js:23487
    callCallback2 @ react-dom.development.js:3945
    invokeGuardedCallbackDev @ react-dom.development.js:3994
    invokeGuardedCallback @ react-dom.development.js:4056
    flushPassiveEffectsImpl @ react-dom.development.js:23574
    unstable_runWithPriority @ scheduler.development.js:468
    runWithPriority$1 @ react-dom.development.js:11276
    flushPassiveEffects @ react-dom.development.js:23447
    performSyncWorkOnRoot @ react-dom.development.js:22269
    (anonymous) @ react-dom.development.js:11327
    unstable_runWithPriority @ scheduler.development.js:468
    runWithPriority$1 @ react-dom.development.js:11276
    flushSyncCallbackQueueImpl @ react-dom.development.js:11322
    flushSyncCallbackQueue @ react-dom.development.js:11309
    scheduleUpdateOnFiber @ react-dom.development.js:21893
    dispatchAction @ react-dom.development.js:16139
    getUsagesFromIndexedDB @ App.tsx:215
    await in getUsagesFromIndexedDB (async)
    (anonymous) @ App.tsx:70
    invokePassiveEffectCreate @ react-dom.development.js:23487
    callCallback2 @ react-dom.development.js:3945
    invokeGuardedCallbackDev @ react-dom.development.js:3994
    invokeGuardedCallback @ react-dom.development.js:4056
    flushPassiveEffectsImpl @ react-dom.development.js:23574
    unstable_runWithPriority @ scheduler.development.js:468
    runWithPriority$1 @ react-dom.development.js:11276
    flushPassiveEffects @ react-dom.development.js:23447
    performSyncWorkOnRoot @ react-dom.development.js:22269
    (anonymous) @ react-dom.development.js:11327
    unstable_runWithPriority @ scheduler.development.js:468
    runWithPriority$1 @ react-dom.development.js:11276
    flushSyncCallbackQueueImpl @ react-dom.development.js:11322
    flushSyncCallbackQueue @ react-dom.development.js:11309
    scheduleUpdateOnFiber @ react-dom.development.js:21893
    dispatchAction @ react-dom.development.js:16139
    getUsagesFromIndexedDB @ App.tsx:215
    await in getUsagesFromIndexedDB (async)
    (anonymous) @ App.tsx:70
    invokePassiveEffectCreate @ react-dom.development.js:23487
    callCallback2 @ react-dom.development.js:3945
    invokeGuardedCallbackDev @ react-dom.development.js:3994
    invokeGuardedCallback @ react-dom.development.js:4056
    flushPassiveEffectsImpl @ react-dom.development.js:23574
    unstable_runWithPriority @ scheduler.development.js:468
    runWithPriority$1 @ react-dom.development.js:11276
    flushPassiveEffects @ react-dom.development.js:23447
    performSyncWorkOnRoot @ react-dom.development.js:22269
    (anonymous) @ react-dom.development.js:11327
    unstable_runWithPriority @ scheduler.development.js:468
    runWithPriority$1 @ react-dom.development.js:11276
    flushSyncCallbackQueueImpl @ react-dom.development.js:11322
    flushSyncCallbackQueue @ react-dom.development.js:11309
    scheduleUpdateOnFiber @ react-dom.development.js:21893
    dispatchAction @ react-dom.development.js:16139
    getUsagesFromIndexedDB @ App.tsx:215
    await in getUsagesFromIndexedDB (async)
    (anonymous) @ App.tsx:70
    invokePassiveEffectCreate @ react-dom.development.js:23487
    callCallback2 @ react-dom.development.js:3945
    invokeGuardedCallbackDev @ react-dom.development.js:3994
    invokeGuardedCallback @ react-dom.development.js:4056
    flushPassiveEffectsImpl @ react-dom.development.js:23574
    unstable_runWithPriority @ scheduler.development.js:468
    runWithPriority$1 @ react-dom.development.js:11276
    flushPassiveEffects @ react-dom.development.js:23447
    performSyncWorkOnRoot @ react-dom.development.js:22269
    (anonymous) @ react-dom.development.js:11327
    unstable_runWithPriority @ scheduler.development.js:468
    runWithPriority$1 @ react-dom.development.js:11276
    flushSyncCallbackQueueImpl @ react-dom.development.js:11322
    flushSyncCallbackQueue @ react-dom.development.js:11309
    scheduleUpdateOnFiber @ react-dom.development.js:21893
    dispatchAction @ react-dom.development.js:16139
    getUsagesFromIndexedDB @ App.tsx:215
    await in getUsagesFromIndexedDB (async)
    (anonymous) @ App.tsx:70
    invokePassiveEffectCreate @ react-dom.development.js:23487
    callCallback2 @ react-dom.development.js:3945
    invokeGuardedCallbackDev @ react-dom.development.js:3994
    invokeGuardedCallback @ react-dom.development.js:4056
    flushPassiveEffectsImpl @ react-dom.development.js:23574
    unstable_runWithPriority @ scheduler.development.js:468
    runWithPriority$1 @ react-dom.development.js:11276
    flushPassiveEffects @ react-dom.development.js:23447
    performSyncWorkOnRoot @ react-dom.development.js:22269
    (anonymous) @ react-dom.development.js:11327
    unstable_runWithPriority @ scheduler.development.js:468
    runWithPriority$1 @ react-dom.development.js:11276
    flushSyncCallbackQueueImpl @ react-dom.development.js:11322
    flushSyncCallbackQueue @ react-dom.development.js:11309
    scheduleUpdateOnFiber @ react-dom.development.js:21893
    dispatchAction @ react-dom.development.js:16139
    getUsagesFromIndexedDB @ App.tsx:215
    await in getUsagesFromIndexedDB (async)
    (anonymous) @ App.tsx:70
    invokePassiveEffectCreate @ react-dom.development.js:23487
    callCallback2 @ react-dom.development.js:3945
    invokeGuardedCallbackDev @ react-dom.development.js:3994
    invokeGuardedCallback @ react-dom.development.js:4056
    flushPassiveEffectsImpl @ react-dom.development.js:23574
    unstable_runWithPriority @ scheduler.development.js:468
    runWithPriority$1 @ react-dom.development.js:11276
    flushPassiveEffects @ react-dom.development.js:23447
    performSyncWorkOnRoot @ react-dom.development.js:22269
    (anonymous) @ react-dom.development.js:11327
    unstable_runWithPriority @ scheduler.development.js:468
    runWithPriority$1 @ react-dom.development.js:11276
    flushSyncCallbackQueueImpl @ react-dom.development.js:11322
    flushSyncCallbackQueue @ react-dom.development.js:11309
    scheduleUpdateOnFiber @ react-dom.development.js:21893
    dispatchAction @ react-dom.development.js:16139
    getUsagesFromIndexedDB @ App.tsx:215
    await in getUsagesFromIndexedDB (async)
    (anonymous) @ App.tsx:70
    invokePassiveEffectCreate @ react-dom.development.js:23487
    callCallback2 @ react-dom.development.js:3945
    invokeGuardedCallbackDev @ react-dom.development.js:3994
    invokeGuardedCallback @ react-dom.development.js:4056
    flushPassiveEffectsImpl @ react-dom.development.js:23574
    unstable_runWithPriority @ scheduler.development.js:468
    runWithPriority$1 @ react-dom.development.js:11276
    flushPassiveEffects @ react-dom.development.js:23447
    performSyncWorkOnRoot @ react-dom.development.js:22269
    (anonymous) @ react-dom.development.js:11327
    unstable_runWithPriority @ scheduler.development.js:468
    runWithPriority$1 @ react-dom.development.js:11276
    flushSyncCallbackQueueImpl @ react-dom.development.js:11322
    flushSyncCallbackQueue @ react-dom.development.js:11309
    scheduleUpdateOnFiber @ react-dom.development.js:21893
    dispatchAction @ react-dom.development.js:16139
    getUsagesFromIndexedDB @ App.tsx:215
    await in getUsagesFromIndexedDB (async)
    (anonymous) @ App.tsx:70
    invokePassiveEffectCreate @ react-dom.development.js:23487
    callCallback2 @ react-dom.development.js:3945
    invokeGuardedCallbackDev @ react-dom.development.js:3994
    invokeGuardedCallback @ react-dom.development.js:4056
    flushPassiveEffectsImpl @ react-dom.development.js:23574
    unstable_runWithPriority @ scheduler.development.js:468
    runWithPriority$1 @ react-dom.development.js:11276
    flushPassiveEffects @ react-dom.development.js:23447
    performSyncWorkOnRoot @ react-dom.development.js:22269
    (anonymous) @ react-dom.development.js:11327
    unstable_runWithPriority @ scheduler.development.js:468
    runWithPriority$1 @ react-dom.development.js:11276
    flushSyncCallbackQueueImpl @ react-dom.development.js:11322
    flushSyncCallbackQueue @ react-dom.development.js:11309
    scheduleUpdateOnFiber @ react-dom.development.js:21893
    dispatchAction @ react-dom.development.js:16139
    getUsagesFromIndexedDB @ App.tsx:215
    await in getUsagesFromIndexedDB (async)
    (anonymous) @ App.tsx:70
    invokePassiveEffectCreate @ react-dom.development.js:23487
    callCallback2 @ react-dom.development.js:3945
    invokeGuardedCallbackDev @ react-dom.development.js:3994
    invokeGuardedCallback @ react-dom.development.js:4056
    flushPassiveEffectsImpl @ react-dom.development.js:23574
    unstable_runWithPriority @ scheduler.development.js:468
    runWithPriority$1 @ react-dom.development.js:11276
    flushPassiveEffects @ react-dom.development.js:23447
    performSyncWorkOnRoot @ react-dom.development.js:22269
    (anonymous) @ react-dom.development.js:11327
    unstable_runWithPriority @ scheduler.development.js:468
    runWithPriority$1 @ react-dom.development.js:11276
    flushSyncCallbackQueueImpl @ react-dom.development.js:11322
    flushSyncCallbackQueue @ react-dom.development.js:11309
    scheduleUpdateOnFiber @ react-dom.development.js:21893
    dispatchAction @ react-dom.development.js:16139
    getUsagesFromIndexedDB @ App.tsx:215
    await in getUsagesFromIndexedDB (async)
    (anonymous) @ App.tsx:70
    invokePassiveEffectCreate @ react-dom.development.js:23487
    callCallback2 @ react-dom.development.js:3945
    invokeGuardedCallbackDev @ react-dom.development.js:3994
    invokeGuardedCallback @ react-dom.development.js:4056
    flushPassiveEffectsImpl @ react-dom.development.js:23574
    unstable_runWithPriority @ scheduler.development.js:468
    runWithPriority$1 @ react-dom.development.js:11276
    flushPassiveEffects @ react-dom.development.js:23447
    performSyncWorkOnRoot @ react-dom.development.js:22269
    (anonymous) @ react-dom.development.js:11327
    unstable_runWithPriority @ scheduler.development.js:468
    runWithPriority$1 @ react-dom.development.js:11276
    flushSyncCallbackQueueImpl @ react-dom.development.js:11322
    flushSyncCallbackQueue @ react-dom.development.js:11309
    scheduleUpdateOnFiber @ react-dom.development.js:21893
    dispatchAction @ react-dom.development.js:16139
    getUsagesFromIndexedDB @ App.tsx:215
    await in getUsagesFromIndexedDB (async)
    (anonymous) @ App.tsx:70
    invokePassiveEffectCreate @ react-dom.development.js:23487
    callCallback2 @ react-dom.development.js:3945
    invokeGuardedCallbackDev @ react-dom.development.js:3994
    invokeGuardedCallback @ react-dom.development.js:4056
    flushPassiveEffectsImpl @ react-dom.development.js:23574
    unstable_runWithPriority @ scheduler.development.js:468
    runWithPriority$1 @ react-dom.development.js:11276
    flushPassiveEffects @ react-dom.development.js:23447
    performSyncWorkOnRoot @ react-dom.development.js:22269
    (anonymous) @ react-dom.development.js:11327
    unstable_runWithPriority @ scheduler.development.js:468
    runWithPriority$1 @ react-dom.development.js:11276
    flushSyncCallbackQueueImpl @ react-dom.development.js:11322
    flushSyncCallbackQueue @ react-dom.development.js:11309
    scheduleUpdateOnFiber @ react-dom.development.js:21893
    dispatchAction @ react-dom.development.js:16139
    getUsagesFromIndexedDB @ App.tsx:215
    await in getUsagesFromIndexedDB (async)
    (anonymous) @ App.tsx:70
    invokePassiveEffectCreate @ react-dom.development.js:23487
    callCallback2 @ react-dom.development.js:3945
    invokeGuardedCallbackDev @ react-dom.development.js:3994
    invokeGuardedCallback @ react-dom.development.js:4056
    flushPassiveEffectsImpl @ react-dom.development.js:23574
    unstable_runWithPriority @ scheduler.development.js:468
    runWithPriority$1 @ react-dom.development.js:11276
    flushPassiveEffects @ react-dom.development.js:23447
    performSyncWorkOnRoot @ react-dom.development.js:22269
    (anonymous) @ react-dom.development.js:11327
    unstable_runWithPriority @ scheduler.development.js:468
    runWithPriority$1 @ react-dom.development.js:11276
    flushSyncCallbackQueueImpl @ react-dom.development.js:11322
    flushSyncCallbackQueue @ react-dom.development.js:11309
    scheduleUpdateOnFiber @ react-dom.development.js:21893
    dispatchAction @ react-dom.development.js:16139
    getUsagesFromIndexedDB @ App.tsx:215
    await in getUsagesFromIndexedDB (async)
    (anonymous) @ App.tsx:70
    invokePassiveEffectCreate @ react-dom.development.js:23487
    callCallback2 @ react-dom.development.js:3945
    invokeGuardedCallbackDev @ react-dom.development.js:3994
    invokeGuardedCallback @ react-dom.development.js:4056
    flushPassiveEffectsImpl @ react-dom.development.js:23574
    unstable_runWithPriority @ scheduler.development.js:468
    runWithPriority$1 @ react-dom.development.js:11276
    flushPassiveEffects @ react-dom.development.js:23447
    performSyncWorkOnRoot @ react-dom.development.js:22269
    (anonymous) @ react-dom.development.js:11327
    unstable_runWithPriority @ scheduler.development.js:468
    runWithPriority$1 @ react-dom.development.js:11276
    flushSyncCallbackQueueImpl @ react-dom.development.js:11322
    flushSyncCallbackQueue @ react-dom.development.js:11309
    scheduleUpdateOnFiber @ react-dom.development.js:21893
    dispatchAction @ react-dom.development.js:16139
    getUsagesFromIndexedDB @ App.tsx:215
    await in getUsagesFromIndexedDB (async)
    (anonymous) @ App.tsx:70
    invokePassiveEffectCreate @ react-dom.development.js:23487
    callCallback2 @ react-dom.development.js:3945
    invokeGuardedCallbackDev @ react-dom.development.js:3994
    invokeGuardedCallback @ react-dom.development.js:4056
    flushPassiveEffectsImpl @ react-dom.development.js:23574
    unstable_runWithPriority @ scheduler.development.js:468
    runWithPriority$1 @ react-dom.development.js:11276
    flushPassiveEffects @ react-dom.development.js:23447
    performSyncWorkOnRoot @ react-dom.development.js:22269
    (anonymous) @ react-dom.development.js:11327
    unstable_runWithPriority @ scheduler.development.js:468
    runWithPriority$1 @ react-dom.development.js:11276
    flushSyncCallbackQueueImpl @ react-dom.development.js:11322
    flushSyncCallbackQueue @ react-dom.development.js:11309
    scheduleUpdateOnFiber @ react-dom.development.js:21893
    dispatchAction @ react-dom.development.js:16139
    getUsagesFromIndexedDB @ App.tsx:215
    await in getUsagesFromIndexedDB (async)
    (anonymous) @ App.tsx:70
    invokePassiveEffectCreate @ react-dom.development.js:23487
    callCallback2 @ react-dom.development.js:3945
    invokeGuardedCallbackDev @ react-dom.development.js:3994
    invokeGuardedCallback @ react-dom.development.js:4056
    flushPassiveEffectsImpl @ react-dom.development.js:23574
    unstable_runWithPriority @ scheduler.development.js:468
    runWithPriority$1 @ react-dom.development.js:11276
    flushPassiveEffects @ react-dom.development.js:23447
    performSyncWorkOnRoot @ react-dom.development.js:22269
    (anonymous) @ react-dom.development.js:11327
    unstable_runWithPriority @ scheduler.development.js:468
    runWithPriority$1 @ react-dom.development.js:11276
    flushSyncCallbackQueueImpl @ react-dom.development.js:11322
    flushSyncCallbackQueue @ react-dom.development.js:11309
    scheduleUpdateOnFiber @ react-dom.development.js:21893
    dispatchAction @ react-dom.development.js:16139
    getUsagesFromIndexedDB @ App.tsx:215
    await in getUsagesFromIndexedDB (async)
    (anonymous) @ App.tsx:70
    invokePassiveEffectCreate @ react-dom.development.js:23487
    callCallback2 @ react-dom.development.js:3945
    invokeGuardedCallbackDev @ react-dom.development.js:3994
    invokeGuardedCallback @ react-dom.development.js:4056
    flushPassiveEffectsImpl @ react-dom.development.js:23574
    unstable_runWithPriority @ scheduler.development.js:468
    runWithPriority$1 @ react-dom.development.js:11276
    flushPassiveEffects @ react-dom.development.js:23447
    performSyncWorkOnRoot @ react-dom.development.js:22269
    (anonymous) @ react-dom.development.js:11327
    unstable_runWithPriority @ scheduler.development.js:468
    runWithPriority$1 @ react-dom.development.js:11276
    flushSyncCallbackQueueImpl @ react-dom.development.js:11322
    flushSyncCallbackQueue @ react-dom.development.js:11309
    batchedUpdates$1 @ react-dom.development.js:22387
    notify2 @ Subscription.js:15
    notifyNestedSubs @ Subscription.js:85
    handleChangeWrapper @ Subscription.js:90
    dispatch @ redux.js:297
    e @ VM8040:2494
    (anonymous) @ serializableStateInvariantMiddleware.ts:206
    (anonymous) @ index.js:20
    (anonymous) @ immutableStateInvariantMiddleware.ts:258
    dispatch @ redux.js:659
    (anonymous) @ createAsyncThunk.ts:649
    step @ index.js:32
    (anonymous) @ index.js:32
    fulfilled @ index.js:32
    Promise.then (async)
    step @ index.js:32
    (anonymous) @ index.js:32
    __async @ index.js:32
    (anonymous) @ createAsyncThunk.ts:580
    (anonymous) @ createAsyncThunk.ts:651
    (anonymous) @ index.js:16
    (anonymous) @ immutableStateInvariantMiddleware.ts:258
    dispatch @ VM8040:3665
    (anonymous) @ App.tsx:77
    invokePassiveEffectCreate @ react-dom.development.js:23487
    callCallback2 @ react-dom.development.js:3945
    invokeGuardedCallbackDev @ react-dom.development.js:3994
    invokeGuardedCallback @ react-dom.development.js:4056
    flushPassiveEffectsImpl @ react-dom.development.js:23574
    unstable_runWithPriority @ scheduler.development.js:468
    runWithPriority$1 @ react-dom.development.js:11276
    flushPassiveEffects @ react-dom.development.js:23447
    performSyncWorkOnRoot @ react-dom.development.js:22269
    (anonymous) @ react-dom.development.js:11327
    unstable_runWithPriority @ scheduler.development.js:468
    runWithPriority$1 @ react-dom.development.js:11276
    flushSyncCallbackQueueImpl @ react-dom.development.js:11322
    flushSyncCallbackQueue @ react-dom.development.js:11309
    unbatchedUpdates @ react-dom.development.js:22438
    legacyRenderSubtreeIntoContainer @ react-dom.development.js:26020
    render @ react-dom.development.js:26103
    (anonymous) @ index.tsx:10
</details>

**Comment by lars-reimann** [comment]
The crash is fixed now. Remaining complaints:

* All zeros should have the minimum color. However, depending on the max value they are colored differently:
image
* The logic that computes the colors should have unit tests.

File: api-editor/gui/src/features/packageData/treeView/TreeNode.tsx [hunk 3]
```diff
@@ -65,4 +109,17 @@ const isSelected = function (declaration: PythonDeclaration, currentPathname: st
     return `/${declaration.pathAsString()}` === currentPathname;
 };
 
+const getColorFromValue = function (maxValue: number, specificValue: number, linear: boolean): string {
+    maxValue++;
+    specificValue++;
+
+    if (specificValue <= 0 || maxValue <= 0) return 'rgb(0%, 0%, 255%)';
```
Comment (lars-reimann) [inline api-editor/gui/src/features/packageData/treeView/TreeNode.tsx:116]: rgb does not use percentages. It should be `rgb(0, 0, 255)`.

**Comment by lars-reimann** [comment]
Looks good to me now.

**Pull request merged**